### ISSUE TITLE
refactor(api): more user friendly error messages

### ISSUE THREAD

**Issue by kszucs** [ibis-project/ibis#6836]
After the switching the validation system to use the new patterns, the messages associated with `ValidationError` exceptions are rather cryptic, containing the the full representation of the (possibly nested) pattern. We need to make these messages more user friendly before the next release.

**Comment by lostmygithubaccount** [comment]
is this something you plan to take on? if not, could you elaborate a bit more on what the current error messages are and where changes are needed? happy to provide help here if I can

**Comment by kszucs** [comment]
Sure, one example is the following:

```py
import ibis

ibis.table({"a":1}, name="e")
```

```py
File ~/Workspace/ibis/ibis/common/annotations.py:78, in Annotation.validate(self, arg, context)
     76 result = self._pattern.match(arg, context)
     77 if result is NoMatch:
---> 78     raise ValidationError(f"{arg!r} doesn't match {self._pattern!r}")
     80 return result

ValidationError: {'a': 1} doesn't match MappingOf(key=InstanceOf(type=<class 'str'>), value=CoercedTo('DataType'), type=Apply(func=<class 'ibis.common.collections.FrozenDict'>))
```

**Comment by cpcloud** [comment]
Closed by #6904
